node/core/approval-voting/src/aux_schema/mod.rs

```diff
@@ -0,0 +1,531 @@
+//! In the future, we may use a temporary DB which doesn't need to be wiped, but for the
+//! time being we share the same DB with the rest of Substrate.
+
+// TODO https://github.com/paritytech/polkadot/1975: remove this
+#![allow(unused)]
+
+use sc_client_api::backend::AuxStore;
+use polkadot_node_primitives::approval::{DelayTranche, RelayVRF};
+use polkadot_primitives::v1::{
+	ValidatorIndex, GroupIndex, CandidateReceipt, SessionIndex, CoreIndex,
+	BlockNumber, Hash, CandidateHash,
+};
+use sp_consensus_slots::SlotNumber;
+use parity_scale_codec::{Encode, Decode};
+
+use std::collections::{BTreeMap, HashMap};
+use std::collections::hash_map::Entry;
+use bitvec::{vec::BitVec, order::Lsb0 as BitOrderLsb0};
+
+use super::Tick;
+
+#[cfg(test)]
+mod tests;
+
+const STORED_BLOCKS_KEY: &[u8] = b"Approvals_StoredBlocks";
+
+/// Metadata regarding a specific tranche of assignments for a specific candidate.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct TrancheEntry {
+	tranche: DelayTranche,
+	// Assigned validators, and the instant we received their assignment, rounded
+	// to the nearest tick.
+	assignments: Vec<(ValidatorIndex, Tick)>,
+}
+
+/// Metadata regarding approval of a particular candidate within the context of some
+/// particular block.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct ApprovalEntry {
+	tranches: Vec<TrancheEntry>,
+	backing_group: GroupIndex,
+	// When the next wakeup for this entry should occur. This is either to
+	// check a no-show or to check if we need to broadcast an assignment.
+	next_wakeup: Tick,
+	our_assignment: Option<OurAssignment>,
+	// `n_validators` bits.
+	assignments: BitVec<BitOrderLsb0, u8>,
+	approved: bool,
+}
+
+/// Metadata regarding approval of a particular candidate.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct CandidateEntry {
+	candidate: CandidateReceipt,
+	session: SessionIndex,
+	// Assignments are based on blocks, so we need to track assignments separately
+	// based on the block we are looking at.
+	block_assignments: BTreeMap<Hash, ApprovalEntry>,
+	approvals: BitVec<BitOrderLsb0, u8>,
+}
+
+/// Metadata regarding approval of a particular block, by way of approval of the
+/// candidates contained within it.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct BlockEntry {
+	block_hash: Hash,
+	session: SessionIndex,
+	slot: SlotNumber,
+	relay_vrf_story: RelayVRF,
+	// The candidates included as-of this block and the index of the core they are
+	// leaving. Sorted ascending by core index.
+	candidates: Vec<(CoreIndex, CandidateHash)>,
+	// A bitfield where the i'th bit corresponds to the i'th candidate in `candidates`.
+	// The i'th bit is `true` iff the candidate has been approved in the context of this
+	// block. The block can be considered approved if the bitfield has all bits set to `true`.
+	approved_bitfield: BitVec<BitOrderLsb0, u8>,
```

Comment:
Actually either all bits for which `self.candidates.binary_search_by(|probe| probe.0.cmp(&i)).is_ok()` holds or else all bits but we set any missing cores initially.

Reply:
This bitfield is by candidates, not by cores. So if there were candidates included from cores `[1, 3, 7]`, then the bit-field would have 3 bits corresponding to each of those, respectively. I think all bits is correct with that in mind, but I might misunderstand the purpose of the binary search snippet you posted if it's not for checking existence of the core in the included candidates set.

It does mean that every time we want to mark a candidate as approved, we have an extra search of the candidate set for each block entry a candidate appears in, but that's contiguous memory and would be dwarfed by the cost of writing that metadata to disk.

Reply:
I only meant that `approved_bitfield` could be initialized to zero for bits representing existent candidates and one for bits representing non-existent candidates.  I just assumed candidates would always be indexed by core in this stage.  Appears you're indexing by some other candidate index, which yes makes `BitVec`'s length is equivalent to presetting non-existent candidates to one.  I see..

Reply:
From the outside of this struct, we do lookup by core, but we have to do an internal look-up via the `candidates` vec first, which is cheap.